### PR TITLE
feat/pdm-lock-automation

### PR DESCRIPTION
**Pull request by kirtimanmishrazipstack**
## What

- Automate pdm.lock upon change in pyproject.toml

## Why

- Every change in `pyproject.toml` required `pdm.lock` to be updated

## How

- Create pdm automate workflow to check for any `uncommited pyproject.toml` change from `main`
- If there is change in `pyproject.toml`, then update pdm.lock in its venv. For eg: backend, prompt-service and so on.

## Can this PR break any existing features. If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)

- NO

## Database Migrations

- NO

## Env Config

- NO

## Relevant Docs

-

## Related Issues or PRs

- https://github.com/Zipstack/unstract/pull/504 This PR needs to be merged before

## Dependencies Versions

-

## Notes on Testing

- Test github workflow. 
- Created a `backend/pyproject.toml` change. pdm lock of backend is updated. 
- `root`, `prompt-service` pdm.lock is `NOT` updated as there is no change in its `pyproject.toml`
- Workflow is successful with 3 matrix. 

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: .github/workflows/pdm-lock-automation.yaml
Comment (ritwik-g) [inline .github/workflows/pdm-lock-automation.yaml:1]: @kirtimanmishrazipstack instead of using reusable workflows can we use matrix? The problem with current approach is that each of this will be executed as separate workflows and also the entire repo is checked out each time I think we can avoid that in this case

File: .github/workflows/reusable-file-change-check.yaml
Comment (ritwik-g) [inline .github/workflows/reusable-file-change-check.yaml:29]: @kirtimanmishrazipstack should we check this against `main` instead of previous HEAD^?

Comment (kirtimanmishrazipstack) [reply .github/workflows/reusable-file-change-check.yaml]: @ritwik-g checked against main now

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml:1]: @ritwik-g using reusble workflow now

File: .pre-commit-config.yaml
Comment (ritwik-g) [inline .pre-commit-config.yaml:101]: @kirtimanmishrazipstack why is this commented?

Comment (kirtimanmishrazipstack) [reply .pre-commit-config.yaml:101]: @ritwik-g github workflow is failing due to this. I think I need to install go like python inside to make it work

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:53]: @kirtimanmishrazipstack might be a good idea to lock this to a version that our Dockerfiles also expect / work with to avoid any surprises

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:37]: Why is this needed?

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:69]:
```suggestion
          pdm lock -G :all
```
This allows us to lock all groups - in case we decide to add deps under a new group, it'll not be picked up automatically

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:89]: When we have separate actions running the precommit checks - why do we run it again? I doubt if the lock files are linted / corrected through these precommit hooks usually

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:94]: @kirtimanmishrazipstack is this condition necessary to commit with different messages? I think we should be fine with just this suggestion right?
```suggestion
            git commit -m "Updated pdm.lock for ${{ matrix.directory }}"
```

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml:89]: @chandrasekharan-zipstack this is done for github runner. Without this, pdmlock commit is not possible

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml:94]: @chandrasekharan-zipstack changed this

Comment (ritwik-g) [reply .github/workflows/pdm-lock-automation.yaml:94]: @kirtimanmishrazipstack I think you haven't pushed this change

File: .github/workflows/pdm-lock-automation.yaml
Comment (ritwik-g) [inline .github/workflows/pdm-lock-automation.yaml:3]: Remove this on push and add trigger for now. Once merged let's test this and make other changes if need. Then we can make this on pull_request.

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:69]:
```suggestion
          pdm lock -G :all -v
```

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:90]:
```suggestion
```
Why is this needed? Do we a make a change in the config file?

**Review by chandrasekharan-zipstack — COMMENTED** [review]
Since the lock version is updated to `4.4.2` and for the sake of correctness - my suggestion is to update the PDM version in our dockerfiles first and then proceed to merge this.
If you've explicitly tested everything we should be fine - but make sure this activity gets tracked as a tech debt

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [reply .github/workflows/pdm-lock-automation.yaml]: Technically its still possible but for the sake of correctness, this is good

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @chandrasekharan-zipstack removed

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @chandrasekharan-zipstack added this as `pdm lock -G :all`

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @ritwik-g done

**Comment by hari-kuriakose** [comment]
@kirtimanmishrazipstack  If PDM version needs to be upgraded for this agent, then we need to upgrade PDM version everywhere else too, in order to avoid surprises. cc @chandrasekharan-zipstack.

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [reply .github/workflows/pdm-lock-automation.yaml]: No harm in making it verbose as well - but we can do that if we face any issues frequently

**Review by chandrasekharan-zipstack — APPROVED** [review]
LGTM - @kirtimanmishrazipstack regarding rolling out this change, as @hari-kuriakose mentioned I prefer updating the pdm version in all our Dockerfiles to `2.16.1` and performing a check with it. Once that PR is merged, we'll be good to merge this PR as well

**Comment by kirtimanmishrazipstack** [comment]
> LGTM - @kirtimanmishrazipstack regarding rolling out this change, as @hari-kuriakose mentioned I prefer updating the pdm version in all our Dockerfiles to `2.16.1` and performing a check with it. Once that PR is merged, we'll be good to merge this PR as well

@chandrasekharan-zipstack @hari-kuriakose please merge pdm version upgrade PR. We can merge this, after pdm version upgrade PR is merged

**Review by ritwik-g — APPROVED** [review]
@kirtimanmishrazipstack shouldn't we remove the backend/pdm.lock file form this PR? Other than that changes looks good to me

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:73]:
```suggestion
          pdm lock -G :all -v
```
Making it verbose to quickly analyze any issues

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:75]:
```suggestion
            echo "No changes between the current branch and the ${{ github.head_ref }} branch."
```

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:79]:
```suggestion
            echo "Changes detected between the current branch and the ${{ github.head_ref }} branch."
```

**Review by chandrasekharan-zipstack — APPROVED** [review]
LGTM

File: .github/workflows/pdm-lock-automation.yaml
Comment (ritwik-g) [reply .github/workflows/pdm-lock-automation.yaml:79]: @chandrasekharan-zipstack isn't the `github.head_ref` the current branch head?

Comment (ritwik-g) [reply .github/workflows/pdm-lock-automation.yaml:79]: May be my understanding is wrong. Raising this to make sure.

Comment (ritwik-g) [reply .github/workflows/pdm-lock-automation.yaml:79]: I would say print both `github.base_ref` and `github.head_ref`.

Check this doc for more details. https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/contexts#github-context

**Review by ritwik-g — APPROVED** [review]
Overall looks good. Check comments from @chandrasekharan-zipstack . 

My one only worry is regarding running in to commit conflict issues. But we will know that only once we start using it. We can fine tune if we face such issues.

File: .github/workflows/pdm-lock-automation.yaml
Comment (gaya3-zipstack) [inline .github/workflows/pdm-lock-automation.yaml:6]: Does this branches indicate the branch against which the PR was raised @kirtimanmishrazipstack ?

Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml:6]: yes @gaya3-zipstack this workflow will run when PR is made against `main`.
Sharing a link for your reference https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml]: done @chandrasekharan-zipstack

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml]: done

File: .github/workflows/pdm-lock-automation.yaml
Comment (kirtimanmishrazipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @ritwik-g ${{ github.head_ref }} is correct.

github.head_ref: This variable refers to the name of the branch where the changes are coming from. In the context of a pull request, it’s the branch that contains the changes you want to merge into the base branch. For example, if you're creating a pull request from a branch named feature-branch into main, github.head_ref would be feature-branch.

github.base_ref: This variable refers to the name of the branch into which the changes are being merged. In the context of a pull request, it's the branch that will receive the changes. Continuing with the previous example, if you're merging feature-branch into main, github.base_ref would be main.

File: .github/workflows/pdm-lock-automation.yaml
Comment (gaya3-zipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @kirtimanmishrazipstack confirmed all issues are sorted out. Hence merging.

Comment (gaya3-zipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @kirtimanmishrazipstack confirmed all issues are sorted out. Hence merging.

File: .github/workflows/pdm-lock-automation.yaml
Comment (gaya3-zipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @kirtimanmishrazipstack confirmed all issues are sorted out. Hence merging.

File: .github/workflows/pdm-lock-automation.yaml
Comment (chandrasekharan-zipstack) [reply .github/workflows/pdm-lock-automation.yaml]: @kirtimanmishrazipstack I agree with @ritwik-g on printing both the base and head ref rather than just going for `current` - it doesn't harm us in any way and it'll only make our lives slightly better while debugging anything on the worfklow

**Comment by chandrasekharan-zipstack** [comment]
@kirtimanmishrazipstack 
1. I understand that going for a matrix of jobs is to run tasks in parallel, but do you have any ideas on how we can encapsulate them all under a single item here? Showing all these checks makes it hard to look into the other checks that run. My suggestion would be to have a separate step which determines whether locking should be done or not and then triggering the matrix of jobs only when there are changes in that service / lib
image
1. Can you mention the service / lib name here? It'll help us distinguish the different runs
image
1. Consider if [this comment](https://github.com/Zipstack/unstract/pull/484#discussion_r1699379655) can be addressed as well - basically printing both the feature branch and the branch we will merge into.

Regarding 1 and 2, let's aim for what's done [here](https://github.com/Zipstack/airbyte/actions/runs/6106650762), the summarize step may not be necessary though - but good to have
image


cc: @ritwik-g @hari-kuriakose

**Comment by kirtimanmishrazipstack** [comment]
1. @chandrasekharan-zipstack I made it as matrix because the alternate to matrix-job is reusable workflow. 
Resuable workflow code will have duplicacy, and also entire repo is checked out each time. @ritwik-g made a comment earlier https://github.com/Zipstack/unstract/pull/484#discussion_r1681131570

3. base ref was printing empty when I tested on act. So I proceed with base ref

**Comment by chandrasekharan-zipstack** [comment]
> 1. @chandrasekharan-zipstack I made it as matrix because the alternate to matrix-job is reusable workflow.
>    Resuable workflow code will have duplicacy, and also entire repo is checked out each time. @ritwik-g made a comment earlier [feat/pdm-lock-automation #484 (comment)](https://github.com/Zipstack/unstract/pull/484#discussion_r1681131570)
> 2. I will go through how to add services and lib, and will raise a PR separately.
> 3. base ref was printing empty when I tested on act. So I proceed with base ref

1. @ritwik-g any ideas on how we can avoid treating each lock automation as a separate check? That's my primary concern
3. @kirtimanmishrazipstack one of `base_ref` or `ref_name` might help - this is anyway a nit concern

**Comment by ritwik-g** [comment]
> > 1. @chandrasekharan-zipstack I made it as matrix because the alternate to matrix-job is reusable workflow.
> >    Resuable workflow code will have duplicacy, and also entire repo is checked out each time. @ritwik-g made a comment earlier [feat/pdm-lock-automation #484 (comment)](https://github.com/Zipstack/unstract/pull/484#discussion_r1681131570)
> > 2. I will go through how to add services and lib, and will raise a PR separately.
> > 3. base ref was printing empty when I tested on act. So I proceed with base ref
> 
> 1. @ritwik-g any ideas on how we can avoid treating each lock automation as a separate check? That's my primary concern
> 2. @kirtimanmishrazipstack one of `base_ref` or `ref_name` might help - this is anyway a nit concern

@chandrasekharan-zipstack regarding point 1. We could use a shell script to do it as single action. I had suggested it but due to time limitations we couldn't crack the same. So that is one improvement we could try. So instead of running this as separate actions we can simply create shell script which can do this

**Pull request closed without merging**
